src/variants/lambda/implementation.mligo

```diff
@@ -0,0 +1,101 @@
+// SPDX-FileCopyrightText: 2022 Tezos Commons
+// SPDX-License-Identifier: LicenseRef-MIT-TC
+
+#if !VARIANT
+#define VARIANT
+
+#include "error_codes.mligo"
+#include "helper/unpack.mligo"
+#include "variants/lambda/types.mligo"
+#include "types.mligo"
+
+// -----------------------------------------------------------------
+// Proposal check
+//
+// This should implement the proposal_check logic of the variant and should
+// verify whether a proposal can be submitted.  It checks 2 things: the
+// proposal itself and the amount of tokens frozen upon submission.  It allows
+// the DAO to reject a proposal by arbitrary logic and captures bond
+// requirements
+// -----------------------------------------------------------------
+
+let handler_exists(handler_name, ce : string * contract_extra) : bool =
+  Big_map.mem handler_name ce.lambdas
+
+let unpack_proposal_metadata (pm: proposal_metadata) : lambda_dao_proposal_metadata =
+  match ((Bytes.unpack pm) : (lambda_dao_proposal_metadata option)) with
+  | Some (v) -> v
+  | None -> (failwith unpacking_proposal_metadata_failed : lambda_dao_proposal_metadata)
+
+let proposal_check (propose_params, ce : propose_params * contract_extra) : unit =
+  match unpack_proposal_metadata(propose_params.proposal_metadata) with
+    | Add_handler ahp -> if handler_exists(ahp.name, ce) then (failwith proposal_handler_exists : unit) else unit
+    | Remove_handler rhp -> if handler_exists(rhp, ce) then unit else (failwith proposal_handler_not_found : unit)
+    | Execute_handler ehp -> (match Big_map.find_opt ehp.handler_name ce.lambdas with
+          | Some hinfo ->
+              let
+                hc_input = (ehp.packed_argument, ce.handler_storage)
+              in hinfo.handler_check hc_input
+          | None -> (failwith proposal_handler_not_found: unit))
+
+let add_handler(ce, ahp : contract_extra * add_handler_param) : contract_extra =
+  if handler_exists(ahp.name, ce)
+    then
+     (failwith proposal_handler_exists: contract_extra)
```

Comment:
Is this not possible to ensure that the handler name is unique during `proposal_check` so that we don't have to check it again in `flush`?

I imagine that if a user submit a proposal with the same handler name to another proposal (whether or not a proposal is ongoing or already completed), we should fail right away and not having to wait until `flush`).

I think we can keep track of `ongoing` proposal's handler name in contract extra, so that we can check it right away in `proposal_check`. Though, this solution require changing `proposal_check` to be able to modify `contract_extra` (adding new handler name if `Add_handler` succeed), not sure if this is worth the effort.

@sras What do you think about this?

Reply:
> think we can keep track of ongoing proposal's handler name in contract extra, so that we can check it right away in proposal_check.

But if I understand your suggestion correctly, what should happen if this ongoing proposal fail? Then our rejection of the new proposal would turn out to be not-required, right?

Either way, I think it is not worth the effort and additional code and associated costs.

Reply:
> But if I understand your suggestion correctly, what should happen if this ongoing proposal fail? Then our rejection of the new proposal would turn out to be not-required, right?

If the ongoing proposal fail, then that would be fine. The reject of the new proposal would make the proposer rename their handler, and proceed to propose again anyway.

If we wait till the proposal is already passed and fail the proposal just because of the name, that would waste the proposer and voters times.

@pasqu4le could give his view on this when he got back.

Reply:
Basically I think users should not forced to select a different name because there is another proposal that proposes to add a handler with same name. Becuase it is just a proposal, and so is the one that is being added. 

I think it makes sense to allow multiple proposals for the same thing, and let only one of them pass.

Reply:
> I think users should not forced to select a different name because there is another proposal that proposes to add a handler with same name. Becuase it is just a proposal, and so is the one that is being added.

I agree, it's only at `flush` that we know if a proposal has been accepted and so that another `handler` with the same name is no longer allowed.

> If we wait till the proposal is already passed and fail the proposal just because of the name, that would waste the proposer and voters times.

That's a valid concern, but a lot of other inefficiencies are saved this way. For example I might want to propose two competing handlers, in which case I could submit both at (roughly) the same time and not have to wait another voting cycle. Likewise this is true even if multiple parties want to propose competing handlers, they don't have to wait until a proposal has been `flush`ed to be able to propose themselves.